File: python_backend/src/arch/command/RecipeStepCommands.py
from uuid import UUID

from pony.orm import db_session, delete

from db import RecipeStep


@db_session
def insert_recipe_step(name: str, number: int, recipe_id: UUID) -> RecipeStep:
    """
    Create a new recipe step
    :return: the new step
    """
    return RecipeStep(step=name, number=number, recipe=recipe_id)


@db_session
def update_recipe_step(recipe_id: UUID, name: str, number: int):
    """
    Update the recipe step
    """
    step = RecipeStep.get(recipe=recipe_id, number=number)
    step.name = name


@db_session
def delete_recipe_step(recipe_id: UUID, number: int):
    step = RecipeStep.get(recipe=recipe_id, number=number)
    delete(step)

File: backend/internal/db/commands/recipeStepCommands.go
package commands

import (
	"github.com/georgysavva/scany/pgxscan"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v4"
	"github.com/viddem/vrecipes/backend/internal/db/tables"
)

var createRecipeStepCommand = `
INSERT INTO recipe_step(recipe_id, number, step, is_heading)
VALUES 				   ($1, 	   $2, 	   $3,   $4)
RETURNING recipe_id, number, step, is_heading
`

func CreateRecipeStep(
	tx pgx.Tx,
	recipeId uuid.UUID,
	number uint16,
	step string,
	isHeading bool,
) (*tables.RecipeStep, error) {
	var recipeStep tables.RecipeStep
	err := pgxscan.Get(
		ctx,
		tx,
		&recipeStep,
		createRecipeStepCommand,
		recipeId,
		number,
		step,
		isHeading,
	)
	return &recipeStep, err
}

var updateRecipeStepCommand = `
UPDATE recipe_step
SET	step=$1, 
	is_heading=$2
WHERE recipe_id=$3 AND number=$4
`

func UpdateRecipeStep(tx pgx.Tx, step string, recipeId uuid.UUID, number uint16, isHeading bool) error {
	_, err := tx.Exec(ctx, updateRecipeStepCommand, step, isHeading, recipeId, number)
	return err
}

var deleteRecipeStepCommand = `
DELETE FROM recipe_step
WHERE recipe_id=$1 AND number=$2
`

func DeleteRecipeStep(tx pgx.Tx, recipeId uuid.UUID, number uint16) error {
	_, err := tx.Exec(ctx, deleteRecipeStepCommand, recipeId, number)
	return err
}
